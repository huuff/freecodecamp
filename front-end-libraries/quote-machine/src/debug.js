import React from "react";
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faArrowUp, faArrowDown } from '@fortawesome/free-solid-svg-icons';

// TODO: The state is only to manage the shown arrow, is there any better way?

export default class Debug extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            hidden: true
        };

        this.changeVisibility = this.changeVisibility.bind(this);
        this.iterableProps = this.iterableProps.bind(this);
    }

    changeVisibility() {
        this.setState((state) => ({
            hidden: !state.hidden
        }))
    }

    iterableProps() {
        return Object.entries(this.props).filter(([prop, val]) => prop !== "logs");
    }

    render() {
        return (
            <div className="card fixed-bottom shadow w-25 ms-3">
                <div className="card-header row align-items-center">
                    <h5 className="col-10">Debug</h5>
                    <button id="toggle-debug" className="float-end col btn" data-bs-toggle="collapse" href="#debug-body" onClick={this.changeVisibility}>
                        <FontAwesomeIcon icon={this.state.hidden ? faArrowUp : faArrowDown}/>
                    </button>
            </div>
            <div id="debug-body" className="collapse">
                <div className="card-body px-3">
                    <ul>
                        {this.iterableProps().map(([prop, val]) => <li key={prop}>{prop}: {val.toString()}</li>)}
                    </ul>
                    <div id="log"> {/* Yes, I know the browser has a console */}
                        <textarea rows="5" id="console" value={this.props.logs.join("\n")} disabled={true} className="w-100 h-75 small"/>
                    </div>
                </div>
            </div>
            </div>
        );
    }
}
