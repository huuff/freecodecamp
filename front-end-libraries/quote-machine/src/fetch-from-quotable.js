export default function fetchFromQuotable(log, params) {
    let url = new URL("http://api.quotable.io/random");
    for (const param in params) {
        url.searchParams.append(param, params[param]);
    }

    log(`Fetching: ${url}`)
    return fetch(url, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json'
        }
    })
        .then(res => res.json())
        .then(res => {
            log(JSON.stringify(res));
            return {
                text: res.content,
                author: res.author,
                authorSlug: res.authorSlug,
                tags: res.tags,
            };
        })
}
