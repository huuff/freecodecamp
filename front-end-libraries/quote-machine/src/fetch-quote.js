export default function fetchQuote(onArrival, log, params) {
    let url = new URL("http://api.quotable.io/random");
    for (const param in params) {
        url.searchParams.append(param, params[param]);
    }

    log(`Fetching: ${url}`)
    const startTime = Date.now();
    fetch(url, {
        method: 'GET',
        headers: {
            'Content-Type': 'application/json'
        }
    })
        .then(res => res.json())
        .then(res => {
          log(JSON.stringify(res));
          return {
            quote: {
                text: res.content,
                author: res.author,
                authorSlug: res.authorSlug,
                tags: res.tags,
            },
            time: Date.now() - startTime
          };
        })
        .then(onArrival);
}
