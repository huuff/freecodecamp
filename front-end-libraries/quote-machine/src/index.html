<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <title>Quote Machine</title>
  </head>
  <body>
    <div id="root" class="container">

    </div>
  </body>
</html>
