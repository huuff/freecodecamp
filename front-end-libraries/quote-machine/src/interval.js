import { createSlice } from '@reduxjs/toolkit'

const AUTO_CHANGE_TIME = 15000;

export default class ChangeQuoteInterval {
    constructor(func, time=AUTO_CHANGE_TIME) {
        this.func = func;
        this.time = time;
        this.intervalId = setInterval(this.func, this.time);
    }

    reset() {
        clearInterval(this.intervalId);
        this.intervalId = setInterval(this.func, this.time);
    }

}

export const fakeInterval = {
    reset() {
        console.log("There's no interval set yet")
    }
}
