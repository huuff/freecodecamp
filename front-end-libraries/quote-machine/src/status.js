import React from 'react'
import { CSSTransition } from 'react-transition-group'
import { useSelector} from 'react-redux'

import store from './store.js'
import {setStatus} from './status-slice'
import {showError} from './visual'

export const STATUS_STRING = {
    OK: "This string shouldn't appear anywhere",
    FETCHED_SAME: "Oops! Seems like I wasn't able to retrieve a quote matching that criteria",
    ERROR: "A generic error occurred",
};

export const changeStatus = (code) => {
        store.dispatch(setStatus({ code: code}))

        if (code !== "OK") {
            showError()
        }
    }

export const StatusAlert = (props) => {

    return (<CSSTransition
                    in={props.showError}
                    timeout={1000}
                    classNames="fade-effect"
                    mountOnEnter={true}
                >
                    <div
                        className="alert alert-danger position-fixed top-0 start-50 translate-middle-x"
                        style={{minWidth: "60%"}}
                    >
                        {STATUS_STRING[props.code]}
                    </div>
            </CSSTransition>)
}
