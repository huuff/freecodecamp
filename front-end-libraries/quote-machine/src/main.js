'use strict';

import React, { useEffect } from "react";
import _ from "lodash";

// Redux actions
import { log } from "./debug"
import { setQuote } from './quote-slice'

import Debug from "./debug.js";
import QuoteBox from "./quote-box.js";
import {StatusAlert, changeStatus} from "./status.js";

import store from './store.js';

import { useSelector } from 'react-redux'
import { showError, showQuote } from './visual';

import ChangeQuoteInterval from './interval'
import { fakeInterval } from './interval'

import { connect } from 'react-redux'

var interval = fakeInterval;

export class Main extends React.Component {
    constructor(props) {
        super(props)

        this.requestQuote = this.requestQuote.bind(this)
        this.changeQuote = this.changeQuote.bind(this)
    }

    componentDidMount() {
        this.requestQuote();
        interval = new ChangeQuoteInterval(() => {
            this.props.log('Triggered interval');
            this.requestQuote();
        })

    }
    changeQuote(response) {
        const previousQuote = store.getState().quote; //get rid of this

        if (_.isEqual(previousQuote, response)) {
            this.props.log("Unable to find quote matching criteria")
            changeStatus("FETCHED_SAME");
        } else {
            changeStatus("OK");
        }

        this.props.setQuote(response)
    }

    requestQuote(params = {}) {
        showQuote()
        setTimeout( () => {
            this.props.fetchQuote(this.props.log, params).then(this.changeQuote)
        }, this.props.waitTime ) //to prevent the quote from being changed while vanishing
        interval.reset();
    }

    render() {
        return (
            <div>
                <StatusAlert
                    code={this.props.status.code}
                    showError={this.props.visual.showError}
                />
                <QuoteBox
                    quote={this.props.quote}
                    showQuote={this.props.visual.showQuote}
                    requestQuote={this.requestQuote} />
                <Debug
                    logs={this.props.debug.logs}
                    changeStatus={this.props.changeStatus}
                    status={this.props.status}
                    visual={this.props.visual}
                />
            </div>
        );
    }
}

const mapStateToProps = (state) => ({
    status: state.status,
    quote: state.quote,
    visual: state.visual,
    debug: state.debug
})

export default connect(mapStateToProps, { log, setQuote })(Main)
