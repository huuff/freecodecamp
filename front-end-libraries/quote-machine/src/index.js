'use strict';

import React, { useEffect } from "react";
import ReactDOM from "react-dom";

import 'bootstrap';
import './main.scss';

import Debug from "./debug.js";
import QuoteBox from "./quote-box.js";
import fetchQuote from "./fetch-quote.js";

const REFRESH_TIME = 1000;
const AUTO_CHANGE_TIME = 15000;

class ChangeInterval {
    constructor(func, time) {
        this.func = func;
        this.time = time;
        this.intervalId = setInterval(this.func, this.time);
    }

    reset() {
        clearInterval(this.intervalId);
        this.intervalId = setInterval(this.func, this.time);
    }

}

class Main extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            quote: {
                text: "",
                author: "",
                authorSlug: "",
                tags: [],
            },
            loading: true,
            interval: new ChangeInterval(() => {
                this.log('Triggered interval');
                this.randomQuote();
            }, AUTO_CHANGE_TIME),
            logs: [],
        };

        this.requestQuote = this.requestQuote.bind(this);
        this.changeQuote = this.changeQuote.bind(this);
        this.log = this.log.bind(this);
    }

    componentDidMount() {
        this.requestQuote({});
    }

    changeQuote(response) {
        setTimeout( () => {
            this.setState(() => ({
                quote: response.quote,
                loading: false
            }));
        }, REFRESH_TIME - response.time);
    }

    requestQuote(params) {
        this.setState((state) => ({
            loading: true,
        }));
        fetchQuote(this.changeQuote, this.log, params);
        this.state.interval.reset();
    }

    log(message) {
        this.setState((state) => ({
            logs: [...state.logs].concat(message)
        }));
    }

    render() {
        return (
            <div>
                <QuoteBox quote={this.state.quote} loading={this.state.loading} requestQuote={this.requestQuote} />
                <Debug loading={this.state.loading} quote={JSON.stringify(this.state.quote)} logs={this.state.logs} interval={this.state.interval} />
            </div>
        );
    }
}


ReactDOM.render(
    <Main />,
    document.getElementById('root')
);

