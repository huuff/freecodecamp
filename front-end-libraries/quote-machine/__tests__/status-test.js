import React from 'react'
import ReactDOM from 'react-dom'
import {StatusAlert} from '../src/status'

// React testing library
import { render, fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'

test('renders correctly', () => {
  const tree = renderer
    .create(<StatusAlert code="ERROR" showError={true}/>)
    .toJSON()

  expect(tree).toMatchSnapshot()
})

