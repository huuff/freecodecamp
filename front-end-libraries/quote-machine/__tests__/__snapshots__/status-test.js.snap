// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="alert alert-danger position-fixed top-0 start-50 translate-middle-x"
  style={
    Object {
      "minWidth": "60%",
    }
  }
>
  A generic error occurred
</div>
`;
